sw/device/lib/dif/dif_spi_host.c

```diff
@@ -0,0 +1,388 @@
+static inline void tx_fifo_write32(const dif_spi_host_t *spi_host,
+                                   uintptr_t srcaddr) {
+  uint32_t *src = (uint32_t *)srcaddr;
+  wait_tx_fifo(spi_host);
+  mmio_region_write32(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET, *src);
+}
+
+void dif_spi_host_fifo_write(const dif_spi_host_t *spi_host, const void *src,
+                             uint16_t len) {
+  uintptr_t ptr = (uintptr_t)src;
+
+  // If the pointer starts mis-aligned, write until we are aligned.
+  while (ptr & 3 && len > 0) {
+    tx_fifo_write8(spi_host, ptr);
+    ptr += 1;
+    len -= 1;
+  }
+
+  // Write complete 32-bit words to the fifo.
+  while (len > 3) {
+    tx_fifo_write32(spi_host, ptr);
+    ptr += 4;
+    len -= 4;
+  }
+
+  // Clean up any leftover bytes.
+  while (len > 0) {
+    tx_fifo_write8(spi_host, ptr);
+    ptr += 1;
+    len -= 1;
+  }
+}
+
+typedef struct queue {
+  int32_t length;
+  union {
+    uint64_t qword;
+    uint32_t word[2];
+    uint8_t byte[8];
+  };
+} queue_t;
+
+static inline void enqueue_byte(queue_t *queue, uint8_t data) {
+  queue->byte[queue->length++] = data;
+}
+
+static inline void enqueue_word(queue_t *queue, uint32_t data) {
+  if (queue->length % sizeof(uint32_t) == 0) {
+    queue->word[queue->length / sizeof(uint32_t)] = data;
+    queue->length += 4;
+  } else {
+    enqueue_byte(queue, (uint8_t)data);
+    data >>= 8;
+    enqueue_byte(queue, (uint8_t)data);
+    data >>= 8;
+    enqueue_byte(queue, (uint8_t)data);
+    data >>= 8;
+    enqueue_byte(queue, (uint8_t)data);
+  }
+}
+
+static inline uint8_t dequeue_byte(queue_t *queue) {
+  uint8_t val = queue->byte[0];
+  queue->qword >>= 8;
+  queue->length -= 1;
+  return val;
+}
+
+static inline uint32_t dequeue_word(queue_t *queue) {
+  uint32_t val = queue->word[0];
+  queue->qword >>= 32;
+  queue->length -= 4;
+  return val;
+}
+
+void dif_spi_host_fifo_read(const dif_spi_host_t *spi_host, void *dst,
+                            uint16_t len) {
+  uintptr_t ptr = (uintptr_t)dst;
+  // We always have to read from the RXFIFO as a 32-bit word.  We use a
+  // two-word queue to handle destination and length mis-alignments.
+  queue_t queue = {
+      0,
+  };
+
+  // If the buffer is misaligned, write a byte at a time until we reach
+  // alignment.
+  while (ptr & 3 && len > 0) {
+    if (queue.length < 1) {
+      wait_rx_fifo(spi_host);
+      enqueue_word(&queue, mmio_region_read32(spi_host->base_addr,
+                                              SPI_HOST_RXDATA_REG_OFFSET));
+    }
+    uint8_t *p = (uint8_t *)ptr;
+    *p = dequeue_byte(&queue);
+    ptr += 1;
+    len -= 1;
+  }
+
+  // While we can write complete words to memory, operate on 4 bytes at a time.
+  while (len > 3) {
+    if (queue.length < 4) {
+      wait_rx_fifo(spi_host);
+      enqueue_word(&queue, mmio_region_read32(spi_host->base_addr,
+                                              SPI_HOST_RXDATA_REG_OFFSET));
+    }
+    uint32_t *p = (uint32_t *)ptr;
```

Comment:
This falls foul of the issue I pointed out elsewhere wrt aliasing.

You need to use `write_32()`.

Reply:
Done.  What about the byte writes elsewhere in this function?

Reply:
Byte writes are ok; `char` (which is what our `uint8_t` always is) is given special dispensation for this purpose.